test/persistence-hooks.suite.js

```diff
@@ -2946,6 +2946,252 @@ module.exports = function(dataSource, should, connectorCapabilities) {
 
     });
 
+    describe('PersistedModel.upsertWithWhere', function() {
+      var createdId = null;
+      var createdExtra = null;
+      var createdName = null;
+
+      it('triggers hooks in the correct order on create', function(done) {
+        monitorHookExecution();
+        TestModel.upsertWithWhere({ where: { extra: 'e0101' }},
+          { id: 'not-found', name: 'not found' },
+          function(err, record, created) {
+            if (err) return done(err);
+            createdId = record.id;
+            createdExtra = record.extra;
+            createdName = record.name;
+            hookMonitor.names.should.eql([
+              'access',
+              'before save',
+              'persist',
+              'loaded',
+              'after save',
+            ]);
+            done();
+          });
+      });
+
+      it('triggers hooks in the correct order on update', function(done) {
+        monitorHookExecution();
+        TestModel.upsertWithWhere({ where: { extra: 'e0101' }},
```

Comment:
Also, while debugging more I found out, if I give some existing instance like other methods have done, it works out as expected :  
  it('triggers hooks in the correct order on update', function(done) {
      monitorHookExecution();
      console.log("existingInstance id:", existingInstance.id);
      TestModel.upsertWithWhere( { id: existingInstance.id},
          { name: 'new name' , extra: 'new extra'},
          function(err, record, created) {
            if (err) return done(err);
            console.log(record);
            hookMonitor.names.should.eql([
              'access',
              'before save',
              'persist',
              'loaded',
              'after save',
            ]);
            done();
          });
    });

    ✓ triggers hooks in the correct order on update
existingInstance id: 1
{ id: '1', name: 'new name', extra: 'new extra' }